go-controller/pkg/libovsdb/util/router.go

```diff
@@ -34,25 +34,47 @@ func CreateDefaultRouteToExternal(nbClient libovsdbclient.Client, clusterRouter,
 	if err != nil {
 		return fmt.Errorf("attempt at finding node gateway router %s network information failed, err: %w", gwRouterName, err)
 	}
-	clusterSubnets := util.GetAllClusterSubnets()
-	for _, subnet := range clusterSubnets {
-		gatewayIP, err := util.MatchFirstIPNetFamily(utilnet.IsIPv6String(subnet.IP.String()), gatewayIPs)
+	clusterSubnetsV4, clusterSubnetsV6 := util.GetClusterSubnetsWithHostPrefix()
+
+	for _, clusterSubnet := range append(clusterSubnetsV4, clusterSubnetsV6...) {
+		isClusterSubnetIPV6 := utilnet.IsIPv6String(clusterSubnet.CIDR.IP.String())
+		gatewayIP, err := util.MatchFirstIPNetFamily(isClusterSubnetIPV6, gatewayIPs)
 		if err != nil {
 			return fmt.Errorf("could not find gateway IP for gateway router %s with family %v: %v", gwRouterName, false, err)
 		}
 		lrsr := nbdb.LogicalRouterStaticRoute{
-			IPPrefix: subnet.String(),
+			IPPrefix: clusterSubnet.CIDR.String(),
 			Nexthop:  gatewayIP.IP.String(),
 			Policy:   &nbdb.LogicalRouterStaticRoutePolicySrcIP,
 		}
+
+		clusterSubnetPrefixLen, _ := clusterSubnet.CIDR.Mask.Size()
 		p := func(lrsr *nbdb.LogicalRouterStaticRoute) bool {
+			// Replace any existing LRSR for the cluster subnet.
+			// Make sure you don't wipe out the existing LRSR via mp0 for the local node subnet
+			// (e.g. 10.244.1.0/24 10.244.1.2 src-ip) and take into account cluster subnet expansion,
+			// which imposes the cluster part of the subnet to stay the same and only allows the mask
+			// to be expanded (i.e. the mask length to be decreased).
+
+			if utilnet.IsIPv6String(lrsr.Nexthop) != isClusterSubnetIPV6 {
+				return false
+			}
+
 			_, itemCIDR, err := net.ParseCIDR(lrsr.IPPrefix)
 			if err != nil {
 				return false
 			}
-			return util.ContainsCIDR(subnet, itemCIDR) &&
-				lrsr.Nexthop == gatewayIP.IP.String() &&
+			itemPrefixLen, _ := itemCIDR.Mask.Size()
+
+			matchSameNodeRoute := util.StrictlyContainsCIDR(clusterSubnet.CIDR, itemCIDR) && // local node subnet is included in cluster subnet
+				itemPrefixLen == clusterSubnet.HostSubnetLength && // host subnet len cannot be modified
 				lrsr.Policy != nil && *lrsr.Policy == nbdb.LogicalRouterStaticRoutePolicySrcIP
+
+			matchClusterSubnetRoute := clusterSubnet.CIDR.IP.Equal(itemCIDR.IP) && // even after expansion, cluster network address cannot change
+				clusterSubnetPrefixLen <= itemPrefixLen && // mask length can only be decreased
+				lrsr.Policy != nil && *lrsr.Policy == nbdb.LogicalRouterStaticRoutePolicySrcIP
+
+			return !matchSameNodeRoute && matchClusterSubnetRoute
```

Comment:
I feel like this could just be:
```
return clusterSubnet.CIDR.IP.Equal(itemCIDR.IP) && clusterSubnetPrefixLen <= itemPrefixLen && itemPrefixLen != clusterSubnet.HostSubnetLength && lrsr.Policy != nil && *lrsr.Policy == nbdb.LogicalRouterStaticRoutePolicySrcIP
```

what am I missing?

Reply:
Aaah. You do have a point. Let me try it out!

Reply:
BTW if we have `10.128.0.0/16/24` as cluster subnet on a regular cluster, what do we do on a 1-node cluster? Can the host prefix be the same as the cluster subnet mask, as in `10.128.0.0/16/16` :-o